adal/src/main/java/com/microsoft/aad/adal/BrokerProxy.java

```diff
@@ -558,34 +567,154 @@ private boolean hasSupportToAddUserThroughBroker(final String brokerPackageName)
+        for (final Signature signature : packageInfo.signatures) {
+            final byte[] rawCert = signature.toByteArray();
+            final InputStream certStream = new ByteArrayInputStream(rawCert);
+
+            final CertificateFactory certificateFactory;
+            final X509Certificate x509Certificate;
+            try {
+                certificateFactory = CertificateFactory.getInstance("X509");
+                x509Certificate = (X509Certificate) certificateFactory.generateCertificate(
+                        certStream);
+                certificates.add(x509Certificate);
+            } catch (final CertificateException e) {
+                throw new AuthenticationException(ADALError.BROKER_APP_VERIFICATION_FAILED);
+            }
+        }
+
+        return certificates;
+    }
+
+    private X509Certificate getSignerCertAndVerifyCertificateChain(
+            final List<X509Certificate> certificates)
+            throws GeneralSecurityException, AuthenticationException {
+        // create certificate chain, find the self signed cert first and chain all the way back
+        // to the signer cert. Also perform certificate signing validation when chaining them back.
+        X509Certificate issuerCert = getSelfSignedCert(certificates);
+        X509Certificate intermediateCert;
+        int count = 1;
+        while (true) {
+            intermediateCert = findCert(issuerCert.getSubjectDN(),
+                    certificates);
+            if (intermediateCert == null) {
+                throw new AuthenticationException(ADALError.BROKER_APP_VERIFICATION_FAILED);
+            }
+
+            //verify certificate is signed with the public key of the issuer certificate
+            intermediateCert.verify(issuerCert.getPublicKey());
+            issuerCert = intermediateCert;
+
+            count++;
+            if (count == certificates.size()) {
+                break;
+            }
+        }
+
+        return intermediateCert;
+    }
+
+    // Will throw if there is more than one self-signed cert found.
+    private X509Certificate getSelfSignedCert(final List<X509Certificate> certs)
+            throws AuthenticationException {
+        int count = 0;
+        X509Certificate selfSignedCert = null;
+        for (final X509Certificate x509Certificate : certs) {
+            if (x509Certificate.getSubjectDN().equals(x509Certificate.getIssuerDN())) {
+                selfSignedCert = x509Certificate;
+                count ++;
+            }
+        }
+
+        if (count > 1) {
+            throw new AuthenticationException(ADALError.BROKER_APP_VERIFICATION_FAILED);
+        }
+
+        return selfSignedCert;
+    }
+
+    final X509Certificate findCert(final Principal issuer, final List<X509Certificate> certs)
+            throws AuthenticationException {
+        X509Certificate intermediateCert = null;
+        int count = 0;
+        for (int i = 0; i < certs.size(); i++) {
```

Comment:
why not use the loop like
for (final X509Certificate x509Certificate : certs) {